src/client/game_client.py

```diff
@@ -0,0 +1,100 @@
+import socket
+import json
+
+from src.constants import Action, Result, SERVER_HOST, SERVER_PORT, MAX_MESSAGE_SIZE
+
+
+class Client:
+    def __init__(self) -> None:
+        self.__socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.__socket.connect((SERVER_HOST, SERVER_PORT))
+
+    def login(self, name: str, password: str = None, game: str = None, num_turns: int = None,
+              num_players: int = None, is_observer: bool = None, is_full: bool = None) -> dict:
+
+        data: dict = {
+            "name": name,
+            "password": password,
+            "game": game,
+            "num_turns": num_turns,
+            "num_players": num_players,
+            "is_observer": is_observer,
+            "is_full": is_full
+        }
+
+        data = {key: value for (key, value) in data.items() if value is not None}
+
+        return self.__send_data(Action.LOGIN, data)
+
+    def logout(self) -> None:
+        self.__send_data(Action.LOGOUT)
+
+    def map(self) -> dict:
+        return self.__send_data(Action.MAP)
+
+    def game_state(self) -> dict:
+        return self.__send_data(Action.GAME_STATE)
+
+    def game_actions(self) -> dict:
+        return self.__send_data(Action.GAME_ACTIONS)
+
+    def turn(self) -> int:
+        try:
+            self.__send_data(Action.TURN)
+        except TimeoutError:
+            return -1
+        else:
+            return 0
+
+    def chat(self, message: str) -> None:
+        self.__send_data(Action.CHAT, {"message": message})
+
+    def move(self, data: dict) -> None:
+        self.__send_data(Action.MOVE, data)
+
+    def shoot(self, data: dict) -> None:
+        self.__send_data(Action.SHOOT, data)
+
+    def disconnect(self) -> None:
+        self.__socket.close()
+
+    @staticmethod
+    def __parse_response(response_msg: bytes) -> [Result, dict]:
+        result_code = int.from_bytes(response_msg[:4], 'little')
+        data_len = int.from_bytes(response_msg[4:8], 'little')
+
+        if data_len == 0:
+            return result_code, None
+
+        data_json = response_msg[8:8 + data_len].decode('utf-8')
+        response_data = json.loads(data_json)
+        return result_code, response_data
+
+    @staticmethod
+    def __encode_message(data: dict) -> bytes:
+        data_json = json.dumps(data).encode('utf-8')
+        data_len = len(data_json)
+        return data_len.to_bytes(4, 'little') + data_json
+
+    def __send_data(self, action: Action, data: dict = None) -> dict:
+        if data is not None:
+            msg = action.value.to_bytes(4, "little") + self.__encode_message(data)
+        else:
+            msg = action.value.to_bytes(8, "little")
+
+        self.__socket.sendall(msg)
+
+        response_msg = self.__socket.recv(MAX_MESSAGE_SIZE)
```

Comment:
Socket.recv can return less data than MAX_MESSAGE_SIZE, you need to subtract data in chunks of 1024/2048 in a loop with a check that all the data was actually subtracted. For this you need to first extract only action code and msg length